File: src/verifier/reputation.py
"""ADR-0004 Decision 3 — the reputation-derivation verifier extension.

Kept apart for the same reason `decision.py` and `allocation.py` are. The
derivations here are deliberately reimplemented from their spec description
(03 § "Consuming reputation, recomputably"), sharing no code with the
TypeScript instance — that mirroring *is* the verifiability guarantee.

Determinism rules, identical on both sides:
- Divergence is relative: integer percent of the estimate, unit-free. An entry
  whose estimated/actual units differ, or whose values are non-integers, or
  whose estimate is not positive, is skipped — never guessed at.
- Decay is exact rational arithmetic over the recency ordering (settlements by
  `published`, ties by digest; per-step decay a ratio of small integers,
  accumulated in arbitrary-precision integers) — never a float exponential.
- Neutral prior: a bidder with no history scores exactly 50.
- `afp:dissentVindicated` is a bonus (+25 on a settled entry; a vindication
  with no usable settled entry counts as accuracy 100).

Per settlement (recency step k back from the newest, weight (num/den)^k):
  a = max(0, 100 - (100·|actual−estimate|) // estimate)   if the entry is usable
  a += dissentBonus                                        if also vindicated
  a = 100                                                  if vindicated only
Score = (Σ a·w) // (Σ w) over the settlements the bidder appears in.
"""

from __future__ import annotations

from decision import instant_millis


def _usable_cost(value) -> tuple[str, int] | None:
    """A cost is usable when it carries a string unit and an **integer-valued
    number**.

    JSON draws no int/float line, and the AFP JCS profile forbids only
    *non-integer* numbers — `100.0` is legal on the wire and canonicalizes to
    `100`, so a signed document may well carry it. Testing `isinstance(x, int)`
    alone would drop such an entry here while the TypeScript writer
    (`Number.isInteger`) keeps it, and the two implementations would score the
    same settlement differently. Accept integer-valued floats, reject booleans
    (a bool is an int in Python, never a cost), reject genuine fractions.
    """
    if not isinstance(value, dict):
        return None
    unit = value.get("unit")
    amount = value.get("value")
    if not isinstance(unit, str) or isinstance(amount, bool):
        return None
    if isinstance(amount, int):
        return unit, amount
    if isinstance(amount, float) and amount.is_integer():
        return unit, int(amount)
    return None


def divergence_decay(params: dict, settlements: list[dict], bidder: str) -> int:
    """settlements: [{"object": <afp:Settlement>, "published": str, "digest": str}]."""
    num = params.get("decayNum", 1)
    den = params.get("decayDen", 2)
    bonus = params.get("dissentBonus", 25)

    ordered = sorted(settlements, key=lambda s: (instant_millis(s["published"]), s["digest"]))
    n = len(ordered)

    score_num = 0
    score_den = 0
    for idx, record in enumerate(ordered):
        settlement = record["object"]
        k = n - 1 - idx  # steps back from the newest
        weight = (num ** k) * (den ** (n - 1 - k))  # common denominator den^(n-1)

        entries = settlement.get("afp:settles") or []
        entry = next((e for e in entries if isinstance(e, dict) and e.get("actor") == bidder), None)
        # Must be a list: a bare `in` against a string is a SUBSTRING test, so
        # a malformed (or crafted) "afp:dissentVindicated": "…sub-bidder-42"
        # would award vindication to bidder "bidder-42" here while the writer,
        # which type-guards with Array.isArray, awards none.
        vindicated_field = settlement.get("afp:dissentVindicated")
        vindicated = isinstance(vindicated_field, list) and bidder in vindicated_field

        a = None
        if entry is not None:
            estimated = _usable_cost((entry.get("afp:estimated") or {}).get("afp:estimatedCost"))
            actual = _usable_cost(entry.get("afp:actual"))
            if estimated and actual and estimated[0] == actual[0] and estimated[1] > 0:
                divergence = (100 * abs(actual[1] - estimated[1])) // estimated[1]
                a = 0 if divergence >= 100 else 100 - divergence
                if vindicated:
                    a += bonus
            elif vindicated:
                a = 100  # unusable entry, but an accurate minority objection still counts
            # Unusable entry, not vindicated: skipped — never guessed at.
        elif vindicated:
            a = 100  # vindication with no settled entry: full accuracy

        if a is not None:
            score_num += a * weight
            score_den += weight

    if score_den == 0:
        return 50  # neutral prior — no history is not bad history
    return score_num // score_den


REPUTATION_RULES = {
    "divergence-decay": divergence_decay,
}

File: src/instance/src/allocation/reputation.ts
/**
 * The reputation-derivation registry (ADR-0004 Decision 3): a small named set
 * of pure functions from a pinned settlement list to a per-bidder score —
 * exactly the selection-rule registry's shape, reapplied. An unknown name is
 * a verification failure, not a skip.
 *
 * Determinism rules (03 § Consuming reputation, recomputably), mirrored
 * bit-for-bit by the Python verifier's `reputation.py`:
 *
 * - Divergence is **relative**: integer percent of the estimate, unit-free.
 *   An entry whose estimated/actual units differ, or whose values are
 *   non-numeric, or whose estimate is not a positive integer, is skipped —
 *   never guessed at.
 * - Decay is **exact rational arithmetic over the recency ordering**:
 *   settlements ordered by `published`, ties by digest; per-step decay a
 *   ratio of small integers (default 1/2), accumulated in bigints — never a
 *   wall-clock float exponential.
 * - Neutral prior: a bidder with no history scores exactly 50 — new entrants
 *   are not punished for being new.
 * - `afp:dissentVindicated` is a bonus (default +25 on a settled entry; a
 *   vindication with no settled entry counts as accuracy 100) — a swarm that
 *   penalizes accurate minority objections stops producing them (04).
 *
 * Per settlement (recency step k back from the newest, weight (num/den)^k):
 *   a = max(0, 100 - floor(100·|actual−estimate| / estimate))   if entry usable
 *   a += dissentBonus                                            if also vindicated
 *   a = 100                                                      if vindicated only
 * Score = floor(Σ a·w / Σ w) over the settlements the bidder appears in.
 */

import type { JsonValue } from "../crypto/jcs.ts";
import { instantMillis } from "../crypto/time.ts";

export interface ReputationRule {
  name: string;
  params: { [key: string]: JsonValue };
}

/** One pinned settlement, resolved from its digest: the afp:Settlement object plus ordering keys. */
export interface PinnedSettlement {
  object: { [key: string]: JsonValue };
  published: string;
  digest: string;
}

type ScoreFn = (params: { [key: string]: JsonValue }, settlements: PinnedSettlement[], bidder: string) => number;

function usableCost(value: JsonValue): { unit: string; value: number } | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  const unit = (value as Record<string, JsonValue>).unit;
  const amount = (value as Record<string, JsonValue>).value;
  if (typeof unit !== "string" || typeof amount !== "number" || !Number.isInteger(amount)) return null;
  return { unit, value: amount };
}

function divergenceDecay(
  params: { [key: string]: JsonValue },
  settlements: PinnedSettlement[],
  bidder: string,
): number {
  const num = BigInt(typeof params.decayNum === "number" ? params.decayNum : 1);
  const den = BigInt(typeof params.decayDen === "number" ? params.decayDen : 2);
  const bonus = BigInt(typeof params.dissentBonus === "number" ? params.dissentBonus : 25);

  // Recency ordering: by published *instant*, ties by digest — the protocol
  // constant. Comparing the raw strings would order a negative-UTC-offset
  // timestamp before a 'Z' one that it actually follows in time, and the
  // verifier must reach the identical ordering (see instant_millis there).
  const ordered = [...settlements].sort((a, b) => {
    const ta = instantMillis(a.published);
    const tb = instantMillis(b.published);
    if (ta !== tb) return ta - tb;
    return a.digest < b.digest ? -1 : a.digest > b.digest ? 1 : 0;
  });
  const n = ordered.length;

  let scoreNum = 0n;
  let scoreDen = 0n;
  for (let idx = 0; idx < n; idx++) {
    const settlement = ordered[idx].object;
    const k = n - 1 - idx; // steps back from the newest
    // Common denominator den^(n-1): weight = num^k · den^(n-1-k), all bigint.
    const weight = num ** BigInt(k) * den ** BigInt(n - 1 - k);

    // Every element is guarded, not just the array: a null or a bare string in
    // `afp:settles` is a property access away from throwing, and a derivation
    // that crashes where the verifier's returns a score has diverged just as
    // surely as one that returns a different number.
    const entries = Array.isArray(settlement["afp:settles"])
      ? (settlement["afp:settles"] as JsonValue[]).filter(
          (e): e is Record<string, JsonValue> => !!e && typeof e === "object" && !Array.isArray(e),
        )
      : [];
    const entry = entries.find((e) => e.actor === bidder);
    const vindicated = Array.isArray(settlement["afp:dissentVindicated"])
      && (settlement["afp:dissentVindicated"] as JsonValue[]).includes(bidder);

    let a: bigint | null = null;
    if (entry) {
      const estimated = usableCost((entry["afp:estimated"] as Record<string, JsonValue> | undefined)?.["afp:estimatedCost"] ?? null);
      const actual = usableCost(entry["afp:actual"] ?? null);
      if (estimated && actual && estimated.unit === actual.unit && estimated.value > 0) {
        const est = BigInt(estimated.value);
        const act = BigInt(actual.value);
        const diff = act > est ? act - est : est - act;
        const divergence = (100n * diff) / est; // floor — integer percent of the estimate
        a = divergence >= 100n ? 0n : 100n - divergence;
        if (vindicated) a += bonus;
      } else if (vindicated) {
        a = 100n; // unusable entry, but an accurate minority objection still counts
      }
      // Unusable entry, not vindicated: skipped — never guessed at.
    } else if (vindicated) {
      a = 100n; // vindication with no settled entry: full accuracy
    }

    if (a !== null) {
      scoreNum += a * weight;
      scoreDen += weight;
    }
  }

  if (scoreDen === 0n) return 50; // neutral prior — no history is not bad history
  return Number(scoreNum / scoreDen); // floor
}

const REGISTRY: Record<string, ScoreFn> = { "divergence-decay": divergenceDecay };

export function knownReputationRule(name: string): boolean {
  return name in REGISTRY;
}

/** Score one bidder under a named derivation. Throws on an unknown name — never silently skips. */
export function runReputationRule(rule: ReputationRule, settlements: PinnedSettlement[], bidder: string): number {
  const fn = REGISTRY[rule.name];
  if (!fn) throw new Error(`unknown reputation rule ${rule.name} — not in the registry`);
  return fn(rule.params, settlements, bidder);
}
